### PR TITLE
Add model and entities

### PR DESCRIPTION
**Pull request by BartoliniAndBorderCollies**
#### Issue
https://trello.com/c/Vx6GoU3Z

### Description
This branch creates the primary application structure - model and entities.

### Motivation
It is a first milestone to develop app.

### Changes introduced by this pull request
- add lombok dependency
- add abstract classes to model
- add enum classes to model
- add entities with fields and relations

### CODE REVIEW

File: src/main/java/com/example/Restaurant_Management_System_REST_API/controller/HomeController.java
Comment (JasinskiSz) [inline src/main/java/com/example/Restaurant_Management_System_REST_API/controller/HomeController.java:1]: I don't understand concept of this controller. It is not even anywhere in the documentation provided as of now (the diagram)

File: src/main/java/com/example/Restaurant_Management_System_REST_API/service/ReservationService.java
Comment (JasinskiSz) [inline src/main/java/com/example/Restaurant_Management_System_REST_API/service/ReservationService.java]: I won't accept anything like that. We talked about it, Bartek. Unimplemented methods are not meant to be on the PRs due to the fact that when those will get on master it will be harder to catch any errors! This is unacceptable.

File: src/main/java/com/example/Restaurant_Management_System_REST_API/model/SingleRecord.java
Comment (JasinskiSz) [inline src/main/java/com/example/Restaurant_Management_System_REST_API/model/SingleRecord.java]: I don't like the name. It's very vague what it represents.

File: src/main/java/com/example/Restaurant_Management_System_REST_API/model/ContactDetails.java
Comment (JasinskiSz) [inline src/main/java/com/example/Restaurant_Management_System_REST_API/model/ContactDetails.java]: Are those fields enough for an address?

File: src/main/java/com/example/Restaurant_Management_System_REST_API/model/entity/Customer.java
Comment (JasinskiSz) [inline src/main/java/com/example/Restaurant_Management_System_REST_API/model/entity/Customer.java]: I'm not sure about this structure where Customer extends ContactDetails. This is somehow a little bit counterintuitive to me why is Customer part of ContactDetails and not the other way around? Why is even one extends another?

**Review by JasinskiSz — CHANGES_REQUESTED** [review]
Why is rest api part doesn't follow a convention of rest api's URL naming? I'm sorry, but I'm not sure how can I accept a structure of this project if there is nothing described here. There is lack of any info what is the purpose of some of those files. To accept it in the current form I'll need:
a) more context, that means more documentation with i.e. UML diagrams and descriptions
b) or at least any sort of description what is what.

I don't know how to give a verdict here and say "yes, this is okay" or "no, this is not okay" as almost everywhere here I'm missing the mentioned before context.

File: src/main/java/com/example/Restaurant_Management_System_REST_API/controller/HomeController.java
Comment (BartoliniAndBorderCollies) [reply src/main/java/com/example/Restaurant_Management_System_REST_API/controller/HomeController.java:1]: this will be a main page of app. 

you already saw it:

image

File: src/main/java/com/example/Restaurant_Management_System_REST_API/service/ReservationService.java
Comment (BartoliniAndBorderCollies) [reply src/main/java/com/example/Restaurant_Management_System_REST_API/service/ReservationService.java]: but I will fulfill these methods in the next step, what is wrong with it? These services implements generic service therefore those methods must be overrided. This is a branch named `app_structure` If I finish these methods here it will no longer be a `app_structure` but the mixture of I dont know what - app structure and partially ready methods in services.

File: src/main/java/com/example/Restaurant_Management_System_REST_API/model/SingleRecord.java
Comment (BartoliniAndBorderCollies) [reply src/main/java/com/example/Restaurant_Management_System_REST_API/model/SingleRecord.java]: it is represents a single record in restaurant menu and in inventory list.  I made it abstract because in both places I need such fields and I didn't want to duplicate code.

File: src/main/java/com/example/Restaurant_Management_System_REST_API/model/ContactDetails.java
Comment (BartoliniAndBorderCollies) [reply src/main/java/com/example/Restaurant_Management_System_REST_API/model/ContactDetails.java]: I think so. Are you proposing anything else?

File: src/main/java/com/example/Restaurant_Management_System_REST_API/model/entity/Customer.java
Comment (BartoliniAndBorderCollies) [reply src/main/java/com/example/Restaurant_Management_System_REST_API/model/entity/Customer.java]: In this context, ContactDetails is an abstract class that holds common information (like name, street, city, postalCode, telephoneNumber, emailAddress) that could be applicable to various entities (like Customer, Supplier, Employee, etc.). So, when Customer extends ContactDetails, it’s inheriting these common properties, indicating that a Customer has these contact details.

If you find this structure confusing, an alternative approach could be to use composition over inheritance. Instead of having Customer extend ContactDetails, I could include ContactDetails as a field within the Customer class.

tell me what you think.

**Comment by BartoliniAndBorderCollies** [comment]
I added comments, please answer me.
ciao!

**Comment by BartoliniAndBorderCollies** [comment]
I deleted empty classes, please review now!

File: src/main/java/com/example/Restaurant_Management_System_REST_API/controller/HomeController.java
Comment (JasinskiSz) [reply src/main/java/com/example/Restaurant_Management_System_REST_API/controller/HomeController.java:1]: I saw it, but then the diagram changed.

File: src/main/java/com/example/Restaurant_Management_System_REST_API/service/ReservationService.java
Comment (JasinskiSz) [reply src/main/java/com/example/Restaurant_Management_System_REST_API/service/ReservationService.java]: The problem is that it'll land on master with these methods returning null. It's unacceptable, as I said previously.

File: src/main/java/com/example/Restaurant_Management_System_REST_API/model/ContactDetails.java
Comment (JasinskiSz) [reply src/main/java/com/example/Restaurant_Management_System_REST_API/model/ContactDetails.java]: I won't suggest anything more at this stage, as I want you to solve it and think :)

**Comment by JasinskiSz** [comment]
> I deleted empty classes, please review now!

It's not ready to review as there are active conversations above. When conversations be resolved you can request my review. Otherwise it's pointless, due to the fact, that I'll review changes I've already reviewed and it's not final version. Please, be cautious of that in the future.

File: src/main/java/com/example/Restaurant_Management_System_REST_API/model/SingleRecord.java
Comment (JasinskiSz) [reply src/main/java/com/example/Restaurant_Management_System_REST_API/model/SingleRecord.java]: Would you be able to find a better name for it? I think there might be better names for that case.

File: src/main/java/com/example/Restaurant_Management_System_REST_API/model/entity/Customer.java
Comment (JasinskiSz) [reply src/main/java/com/example/Restaurant_Management_System_REST_API/model/entity/Customer.java]: That's right, I think inheritance here in this context is not fully correct and/or clear. The other option you described here is more convenient or at last more clear to my with its purpose.

Comment (JasinskiSz) [reply src/main/java/com/example/Restaurant_Management_System_REST_API/model/entity/Customer.java]: And this comment above is my point exactly why requesting review before I was able to answer here is a mistake. That's because I have nothing to review due to the changes that you will now most probably implement. That's FYI to my comment on this matter.

**Comment by BartoliniAndBorderCollies** [comment]
In previous projects, you advised me not to close conversations. Many of them remained unresolved. You also conducted reviews and even accepted pull requests. Please, be consistent.

File: src/main/java/com/example/Restaurant_Management_System_REST_API/model/entity/Customer.java
Comment (BartoliniAndBorderCollies) [reply src/main/java/com/example/Restaurant_Management_System_REST_API/model/entity/Customer.java]: @JasinskiSz I didn't get your answer here. please answer

**Comment by JasinskiSz** [comment]
It depends on the conversations, Bartek :) If there is a convo about something that needs to be changed and you provide me with solution but it's not implemented in code yet and then asking for review, what is the purpose of asking? I'll have more work, because I'll have to request changes again and describe them.

File: src/main/java/com/example/Restaurant_Management_System_REST_API/model/entity/Customer.java
Comment (JasinskiSz) [reply src/main/java/com/example/Restaurant_Management_System_REST_API/model/entity/Customer.java]: @BartoliniAndBorderCollies What's the question I should answer here?

File: src/main/java/com/example/Restaurant_Management_System_REST_API/model/SingleRecord.java
Comment (BartoliniAndBorderCollies) [reply src/main/java/com/example/Restaurant_Management_System_REST_API/model/SingleRecord.java]: Maybe you can propose something? I don't want to guess what you have in your mind and I don't want to play in guessing game like "or maybe this?  not? so maybe this? or maybe this? ". 
I just can tell you that `Record` cannot be -> IntelliJ takes it as a interior part of Java class.
What I could propose else than SingleRecord is `CatalogItem`.
if you have other concept for this name please share with me

File: src/main/java/com/example/Restaurant_Management_System_REST_API/model/entity/Customer.java
Comment (BartoliniAndBorderCollies) [reply src/main/java/com/example/Restaurant_Management_System_REST_API/model/entity/Customer.java]: Ok sorry your answers haven't appeared earlier to me, now I can see them.
thanks!

File: src/main/java/com/example/Restaurant_Management_System_REST_API/model/SingleRecord.java
Comment (JasinskiSz) [reply src/main/java/com/example/Restaurant_Management_System_REST_API/model/SingleRecord.java]: Please, don't take it like that. It's not a guessing game and I don't have anything specific in my mind. I just asked you to think of a better name, because this one is vague and should be changed.
Should I provide you with a list of names which could be correct in this case? I think it's not a learning path I'd like to go with, as I want you to be a standalone, mature developer.

`CatalogItem` is a better name as it represents something and could be easily extended by inventory and menu item. Why is it better? Simply because of the "item" in the name. SingleRecord means something and nothing at the same time. And when other items (inventory and menu) are extending it, it's just confusing I'd say, at least for me.
I'm thinking if there is even a better name to it now... We should consider what this class is all about and maybe come with ideas from here. This is approvable but if I'd find better idea, I'd change the name later.

Comment (BartoliniAndBorderCollies) [reply src/main/java/com/example/Restaurant_Management_System_REST_API/model/SingleRecord.java]: I just asked if you have anything in your mind cause it sounded like that :)
Ok, so I rename it to CatalogItem for now and later we will see!

**Comment by BartoliniAndBorderCollies** [comment]
I resolved all conversations and made appropriate changes. I hope it is ready for re-review -> but still finger is afraid to click it ;P

File: src/main/java/com/example/Restaurant_Management_System_REST_API/controller/HomeController.java
Comment (JasinskiSz) [reply src/main/java/com/example/Restaurant_Management_System_REST_API/controller/HomeController.java:1]: I need further explanation here. What do you mean by "main page" of the app?

File: src/main/java/com/example/Restaurant_Management_System_REST_API/model/entity/Customer.java
Comment (JasinskiSz) [inline src/main/java/com/example/Restaurant_Management_System_REST_API/model/entity/Customer.java:32]: Could you describe for me this 21st line of code? What purpose and what goal it achieves?

File: src/main/java/com/example/Restaurant_Management_System_REST_API/model/entity/MenuRecord.java
Comment (JasinskiSz) [inline src/main/java/com/example/Restaurant_Management_System_REST_API/model/entity/MenuRecord.java]: I think there might be a problem with this type. Is List enough here? Convince me! :D

**Review by JasinskiSz — COMMENTED** [review]
I didn't reviewed all the code. I'll do it tomorrow.

File: src/main/java/com/example/Restaurant_Management_System_REST_API/model/entity/Customer.java
Comment (BartoliniAndBorderCollies) [reply src/main/java/com/example/Restaurant_Management_System_REST_API/model/entity/Customer.java:32]: There will be a role CUSTOMER so in the database personnel can check if a specific customer has any reservation

File: src/main/java/com/example/Restaurant_Management_System_REST_API/model/entity/MenuRecord.java
Comment (BartoliniAndBorderCollies) [reply src/main/java/com/example/Restaurant_Management_System_REST_API/model/entity/MenuRecord.java]: I haven't thought about this!  But when I think I think it is good idea to change it to Set! I will not have duplicate elements, I don't need an insertion order, I will not have null values!
Then I change to Set. is that fair enough?

File: src/main/java/com/example/Restaurant_Management_System_REST_API/model/entity/Customer.java
Comment (BartoliniAndBorderCollies) [reply src/main/java/com/example/Restaurant_Management_System_REST_API/model/entity/Customer.java:32]: ps. I know it is not on diagram, I tried to update it but I cannot reopen it on draw.io -> I got the error and I don't want to do this diagram again!!!!

File: src/main/java/com/example/Restaurant_Management_System_REST_API/controller/HomeController.java
Comment (BartoliniAndBorderCollies) [reply src/main/java/com/example/Restaurant_Management_System_REST_API/controller/HomeController.java:1]: this will be a main page on frontend with nice description of the app, photos and hyperlinks depending what you want to do.

File: src/main/java/com/example/Restaurant_Management_System_REST_API/model/entity/MenuRecord.java
Comment (JasinskiSz) [reply src/main/java/com/example/Restaurant_Management_System_REST_API/model/entity/MenuRecord.java]: Not sure if any of those what you mentioned is a problem here :) Stick with list or change to Set, whichever you prefer then. I'll come back to this conversation in future PR.

File: src/main/java/com/example/Restaurant_Management_System_REST_API/model/entity/Customer.java
Comment (JasinskiSz) [reply src/main/java/com/example/Restaurant_Management_System_REST_API/model/entity/Customer.java:32]: Okay, I'll leave it as it is, but I can see an issue here in future development, and I didn't mean any diagram.
If this occur, I'll come back to this conversation. Feel free to resolve this convo.

File: src/main/java/com/example/Restaurant_Management_System_REST_API/model/entity/Order.java
Comment (JasinskiSz) [inline src/main/java/com/example/Restaurant_Management_System_REST_API/model/entity/Order.java]: Why would you want to use specifically `ArrayList` here?

```suggestion
    private List<MenuRecord> menuRecords;
```

File: src/main/java/com/example/Restaurant_Management_System_REST_API/model/entity/Supplier.java
Comment (JasinskiSz) [inline src/main/java/com/example/Restaurant_Management_System_REST_API/model/entity/Supplier.java]: Same issue here.

**Review by JasinskiSz — COMMENTED** [review]
I have a problem with checking this PR. I'm not sure what some of the fields are doing here but I don't want to write everywhere a comment with request for explanation. I'll take as an example `deliveryDate` in `InventoryItem` which imo is not usable in any way but maybe you see some purpose?

That's my problem with this PR, there is more of those things throughout the classes and I can't compare it to documentation which would describe clearly what is where and why. I think you should build your application further and then ask for my review on this matter as for now I can't do much but just to ask you "why is xyz here?".

Taking into account the above, I think a good idea would be if I just check for some critical issues (like this ArrayList in this PR) and let you develop everything further. 

My recommendation though would be to have another branch named "develop" which will act just like "master" (or "main") but it is meant for development and further testing code. That means that PRs should merge with `dev` branch and after some merges check all the code together and then draw the conclusion if everything looks right.

File: src/main/java/com/example/Restaurant_Management_System_REST_API/controller/HomeController.java
Comment (JasinskiSz) [reply src/main/java/com/example/Restaurant_Management_System_REST_API/controller/HomeController.java:1]: But the client wanted a REST API backend service. If you want to create a frontend for it in the monolith app, then only a "HomeController" is not enough. I'd suggest leaving frontend page for the last, optional step, as it'll make the project even more complex, especially without planning it with diagrams :)

Comment (JasinskiSz) [reply src/main/java/com/example/Restaurant_Management_System_REST_API/controller/HomeController.java:1]: Let me know what you decide.

Comment (BartoliniAndBorderCollies) [reply src/main/java/com/example/Restaurant_Management_System_REST_API/controller/HomeController.java:1]: aaa okay! I understand! then we leave it for now! thanks!

File: src/main/java/com/example/Restaurant_Management_System_REST_API/model/entity/Order.java
Comment (BartoliniAndBorderCollies) [reply src/main/java/com/example/Restaurant_Management_System_REST_API/model/entity/Order.java]: my mistake! it should be List because I want my program more generic.
thanks!

File: src/main/java/com/example/Restaurant_Management_System_REST_API/model/entity/Supplier.java
Comment (BartoliniAndBorderCollies) [reply src/main/java/com/example/Restaurant_Management_System_REST_API/model/entity/Supplier.java]: it should be List definitely!

**Comment by BartoliniAndBorderCollies** [comment]
I understand your difficulties.
About develop branch -> from which point it should begin? I already merged spring security branch to master. 
what about this current branch? I corrected ArrayLists to Lists, is that all for now? shall it be then merged to master or already to develop?
Should develop be a clone of current master or just totally clean?
let me know!
cheers

**Comment by JasinskiSz** [comment]
> About develop branch -> from which point it should begin? I already merged spring security branch to master. 

It could start from latest Master.

> what about this current branch? I corrected ArrayLists to Lists, is that all for now? shall it be then merged to master or already to develop? 

I don't know yet, I'll have to review changes.

**Review by JasinskiSz — COMMENTED** [review]
I no longer see any critical issues. I still don't know if the structure is correct and where will the project go further with those and what problems it'll face, so I didn't paid a lot of attention to entity classes and their "logical" structure. 
And I'm not explicitly accepting this PR. You should decide what to do now, to merge or plan accordingly. My job on this PR looks like done :)

**Comment by BartoliniAndBorderCollies** [comment]
Ok! thanks for your review and all your comments and suggestions.
I created develop branch to which from now I will merge next branches. When you say it is ready I will merge everything to master.
For this branch I believe this structure is well done according to my plan, therefore I merge it to develop.
thanks and cheers!

**Comment by BartoliniAndBorderCollies** [comment]
during a merging I got info some data can be lost so I decide to merge it to master straight ahead and next branches add to develop

**Pull request merged**
